[VL] LATERAL VIEW OUTER stack aborts columnar hash shuffle: "Partition id (field 0) should be integer, but got BOOLEAN"

### Backend

VL (Velox)

### Bug description

For an **OUTER** generator, Velox's `Unnest` operator appends a trailing `BOOLEAN` marker column to its output (output column order: replicated columns → unnest value columns → `[ordinality]` → `[marker]`). Gluten's `GenerateExecTransformer` must consume that marker in `pullOutPostProject` so the native output aligns positionally with the declared Spark schema.

`pullOutPostProject` handles the OUTER marker for `PosExplode`, `Inline`/`JsonTupleExplode`, and `Explode` — but **not for `Stack`**. An OUTER `Stack` falls through to `case _ => generate`, so the marker column is never consumed. Because Velox binds columns **by position**, the unconsumed marker makes the native output one column wider than Gluten's declared schema, shifting every upstream column by one index.

The shift is only observable when a **columnar hash exchange** consumes the Generate output. Under columnar shuffle, `VeloxShuffleWriter::getFirstColumn` requires field 0 of the shuffle `RowVector` to be the int32 `hash_partition_key`. With the columns shifted, the boolean marker lands at field 0 and every task of the stage aborts deterministically:

```
Partition id (field 0) should be integer, but got BOOLEAN
  (cpp/velox/shuffle/VeloxShuffleWriter.h — getFirstColumn)
```

The same unconsumed-marker root cause can surface under different messages depending on the operator downstream of the shift:
- **hash-partition shuffle → SortMergeJoin**: `Partition id (field 0) should be integer, but got BOOLEAN` (the case above).
- **partial HashAggregate** (e.g. `GROUP BY` on the exploded columns): crashes earlier in the native input stream with `values_->capacity() >= byteSize` (`FlatVector.h`).
- **BroadcastHashJoin** downstream: `Serialized encoding is not compatible ... Expected INT_ARRAY. Got BYTE_ARRAY` (`checkTypeEncoding`).

Inner (non-OUTER) `stack` emits no marker column and is unaffected. The existing `test stack function` only exercises inner `stack`, which is why the gap was not caught.

Self-contained repro (crashes before the fix, returns correct rows after):

```sql
CREATE OR REPLACE TEMPORARY VIEW t1_stack AS
  SELECT * FROM VALUES (1,'james',10,'lucy'),(2,'bond',20,'lily') AS tbl(id,name,id1,name1);
CREATE OR REPLACE TEMPORARY VIEW t2_dim AS
  SELECT * FROM VALUES (1,'a'),(2,'b'),(10,'c'),(20,'d') AS tbl(k,tag);

SET spark.sql.autoBroadcastJoinThreshold=-1;   -- force SortMergeJoin (hash-partition columnar exchange)

SELECT j.eq_pos, t2.tag
FROM (
  SELECT eq_pos, val
  FROM t1_stack
  LATERAL VIEW OUTER stack(2, id, name, id1, name1) v AS eq_pos, val
) j
JOIN t2_dim t2 ON j.eq_pos = t2.k
ORDER BY j.eq_pos, t2.tag;
```

Expected (after fix): `(1,a) (2,b) (10,c) (20,d)`.

Root cause is in `PullOutGenerateProjectHelper.pullOutPostProject`
(`backends-velox/src/main/scala/org/apache/gluten/execution/GenerateExecTransformer.scala`):
`Stack` has no `if generate.outer` branch, unlike `Explode`/`PosExplode`/`Inline`.

### Gluten version

main branch

### Spark version

Spark-3.5.x

### Spark configurations

Reproduces on the columnar shuffle path. Force a SortMergeJoin so the exploded key drives a hash-partition columnar exchange:

```
spark.sql.autoBroadcastJoinThreshold = -1
```

### System information

Velox backend. Not environment-specific — the crash is deterministic and data-independent (every task of the stage fails).

### Relevant logs

```bash
Caused by: org.apache.gluten.exception.GlutenException: Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Partition id (field 0) should be integer, but got BOOLEAN
Retriable: False
Expression: firstChild->type()->isInteger()
Function: getFirstColumn
File: /home/ngadm/builds/2Grm8K_1/0/exa/gluten/cpp/velox/shuffle/VeloxShuffleWriter.h
Line: 80
Stack trace:
# 0  
# 1  
# 2  
# 3  
# 4  
# 5  
# 6  
# 7  
# 8  
# 9  
# 10 
# 11 
# 12 
# 13 
# 14 
# 15 
# 16 
# 17 
# 18 
# 19 
# 20 
# 21 
# 22 
# 23 
# 24 
# 25 
# 26 
# 27 
# 28 
# 29 
# 30 
# 31 
# 32 
"
```

--- a/backends-velox/src/main/scala/org/apache/gluten/execution/GenerateExecTransformer.scala
+++ b/backends-velox/src/main/scala/org/apache/gluten/execution/GenerateExecTransformer.scala
@@ -333,33 +333,13 @@ object PullOutGenerateProjectHelper extends PullOutProjectHelper {
             }
             ProjectExec(generate.requiredChildOutput ++ newOutput, newGenerate)
           }
-        case Explode(_) if generate.outer =>
-          // Drop the last column of generatorOutput, which is the boolean representing whether
-          // the null value is unnested from the input array/map (e.g. array(1, null)), or the
-          // array/map itself is null or empty (e.g. array(), map(), null).
-          val isPresent =
-            AttributeReference(generatePostAliasName, BooleanType, nullable = true)()
-          val newGenerate =
-            generate.copy(generatorOutput = generate.generatorOutput :+ isPresent)
-          val newOutput = generate.generatorOutput.map {
-            attr =>
-              val caseWhen = CaseWhen(
-                Seq((isPresent, attr)),
-                Literal(null, attr.dataType)
-              )
-              Alias(caseWhen, generatePostAliasName)(attr.exprId, attr.qualifier)
-          }
-          ProjectExec(generate.requiredChildOutput ++ newOutput, newGenerate)
-        case _: Stack if generate.outer =>
-          // Like the outer Explode above, Velox's Unnest appends a trailing boolean marker
-          // column for an OUTER generator (true when the row is a real unnested value, false
-          // for the synthetic padding row of an empty/null input). Stack emits no ordinality
-          // column, so its native output layout (value columns + trailing marker) matches
-          // Explode's. Without consuming the marker here, the native output is one column wider
-          // than the declared schema, every upstream column shifts by one, and a downstream
-          // hash shuffle receives the boolean marker at field 0 instead of the int32 partition
-          // key (VeloxShuffleWriter getFirstColumn aborts). Append a boolean isPresent to absorb
-          // the marker and wrap each output column so the padding row projects NULLs.
+        case (_: Explode | _: Stack) if generate.outer =>
+          // Drop the last column of generatorOutput, which is the boolean marker Velox's Unnest
+          // appends for an OUTER generator (true when the row is a real unnested value, false for
+          // the synthetic padding row of an empty/null input). Explode and Stack share this
+          // layout -- value columns followed by the trailing marker, with no ordinality column --
+          // so the same handling applies to both. Wrap each output column in a CaseWhen on the
+          // marker so the padding row projects NULLs.
           val isPresent =
             AttributeReference(generatePostAliasName, BooleanType, nullable = true)()
           val newGenerate =

--- a/backends-velox/src/main/scala/org/apache/gluten/execution/GenerateExecTransformer.scala
+++ b/backends-velox/src/main/scala/org/apache/gluten/execution/GenerateExecTransformer.scala
@@ -350,6 +350,29 @@ object PullOutGenerateProjectHelper extends PullOutProjectHelper {
               Alias(caseWhen, generatePostAliasName)(attr.exprId, attr.qualifier)
           }
           ProjectExec(generate.requiredChildOutput ++ newOutput, newGenerate)
+        case _: Stack if generate.outer =>
+          // Like the outer Explode above, Velox's Unnest appends a trailing boolean marker
+          // column for an OUTER generator (true when the row is a real unnested value, false
+          // for the synthetic padding row of an empty/null input). Stack emits no ordinality
+          // column, so its native output layout (value columns + trailing marker) matches
+          // Explode's. Without consuming the marker here, the native output is one column wider
+          // than the declared schema, every upstream column shifts by one, and a downstream
+          // hash shuffle receives the boolean marker at field 0 instead of the int32 partition
+          // key (VeloxShuffleWriter getFirstColumn aborts). Append a boolean isPresent to absorb
+          // the marker and wrap each output column so the padding row projects NULLs.
+          val isPresent =
+            AttributeReference(generatePostAliasName, BooleanType, nullable = true)()
+          val newGenerate =
+            generate.copy(generatorOutput = generate.generatorOutput :+ isPresent)
+          val newOutput = generate.generatorOutput.map {
+            attr =>
+              val caseWhen = CaseWhen(
+                Seq((isPresent, attr)),
+                Literal(null, attr.dataType)
+              )
+              Alias(caseWhen, generatePostAliasName)(attr.exprId, attr.qualifier)
+          }
+          ProjectExec(generate.requiredChildOutput ++ newOutput, newGenerate)
         case _ => generate
       }
     } else {

--- a/backends-velox/src/test/scala/org/apache/gluten/execution/MiscOperatorSuite.scala
+++ b/backends-velox/src/test/scala/org/apache/gluten/execution/MiscOperatorSuite.scala
@@ -1092,6 +1092,50 @@ class MiscOperatorSuite extends VeloxWholeStageTransformerSuite with AdaptiveSpa
     }
   }
 
+  test("test LATERAL VIEW OUTER stack followed by hash shuffle") {
+    // Repro for: `Partition id (field 0) should be integer, but got BOOLEAN`
+    // (VeloxShuffleWriter.h getFirstColumn).
+    //
+    // With OUTER, Velox's Unnest appends a trailing BOOLEAN marker column. The Stack path has
+    // no pullOutPostProject branch to consume that marker (unlike explode/posexplode/inline),
+    // so the native output is one column wider than the declared schema and every upstream
+    // column shifts by one. When the exploded key drives a hash-partition shuffle, the int32
+    // hash_partition_key that must sit at field 0 is displaced by the boolean marker and the
+    // columnar shuffle writer aborts.
+    //
+    // The shuffle must be a *hash-partition* exchange feeding a SortMergeJoin (not a partial
+    // aggregate, not a broadcast join) to reproduce the field-0 crash exactly as production
+    // does: broadcasting the dim table hits a different serializer error, and inserting a
+    // partial HashAggregate crashes earlier in the native input stream. Disable broadcast to
+    // force the SortMergeJoin.
+    withTempView("t1_stack", "t2_dim") {
+      sql("""SELECT * from values
+            |  (1, "james", 10, "lucy"),
+            |  (2, "bond", 20, "lily")
+            |as tbl(id, name, id1, name1)
+         """.stripMargin).createOrReplaceTempView("t1_stack")
+      sql("""SELECT * from values
+            |  (1, "a"), (2, "b"), (10, "c"), (20, "d")
+            |as tbl(k, tag)
+         """.stripMargin).createOrReplaceTempView("t2_dim")
+
+      withSQLConf("spark.sql.autoBroadcastJoinThreshold" -> "-1") {
+        runQueryAndCompare(s"""
+                              |SELECT j.eq_pos, t2.tag
+                              |FROM (
+                              |  SELECT eq_pos, val
+                              |  FROM t1_stack
+                              |  LATERAL VIEW OUTER stack(2, id, name, id1, name1) v AS eq_pos, val
+                              |) j
+                              |JOIN t2_dim t2 ON j.eq_pos = t2.k
+                              |ORDER BY j.eq_pos, t2.tag
+                              |""".stripMargin) {
+          checkGlutenPlan[GenerateExecTransformer]
+        }
+      }
+    }
+  }
+
   test("test inline function") {
     Seq(true, false).foreach {
       isOuter =>

--- a/backends-velox/src/test/scala/org/apache/gluten/execution/MiscOperatorSuite.scala
+++ b/backends-velox/src/test/scala/org/apache/gluten/execution/MiscOperatorSuite.scala
@@ -1092,10 +1092,7 @@ class MiscOperatorSuite extends VeloxWholeStageTransformerSuite with AdaptiveSpa
     }
   }
 
-  test("test LATERAL VIEW OUTER stack followed by hash shuffle") {
-    // Repro for: `Partition id (field 0) should be integer, but got BOOLEAN`
-    // (VeloxShuffleWriter.h getFirstColumn).
-    //
+  test("LATERAL VIEW OUTER stack followed by hash shuffle") {
     // With OUTER, Velox's Unnest appends a trailing BOOLEAN marker column. The Stack path has
     // no pullOutPostProject branch to consume that marker (unlike explode/posexplode/inline),
     // so the native output is one column wider than the declared schema and every upstream

